### PR TITLE
Optimise QubitPermutationGate decomposition

### PR DESCRIPTION
**Pull request by migueltorrescosta**
Solves https://github.com/quantumlib/Cirq/issues/5097 as per comment https://github.com/quantumlib/Cirq/issues/5097#issuecomment-2097041496 .

### CODE REVIEW

File: cirq-core/cirq/ops/permutation_gate.py [hunk 1]
```diff
@@ -74,23 +74,14 @@ def _has_unitary_(self):
...
-        while not is_sorted:
-            is_sorted = True
-            for i in range(0, n - 1, 2):
-                yield from _swap_if_out_of_order(i)
-            for i in range(1, n - 1, 2):
-                yield from _swap_if_out_of_order(i)
+        reversed_permutation_map = {v: i for i, v in enumerate(self.permutation)}
+
+        while reversed_permutation_map:
+            a = list(reversed_permutation_map.keys())[0]
```
Comment (NoureldinYosri) [inline cirq-core/cirq/ops/permutation_gate.py:80]: This code has time complexity O(num_qubits * num_cycles) but can be done in O(num_qubits), try

```py3
permutation = self.permutation.copy()

for i in range(len(permutation)):
   if permutation[i] = -1: continue
   cycle = [i]
   while permutation[cycle[-1]] != i:
      cycle.append(permutation[cylce[-1]])
   for j in cycle: permutation[j] = -1
   cycle.reverse() 
   # yield len(cycle) - 1 swap operations
   


```

**Review by NoureldinYosri — CHANGES_REQUESTED** [review]

File: cirq-core/cirq/ops/permutation_gate.py (hunk 1 above)
Comment (NoureldinYosri) [reply cirq-core/cirq/ops/permutation_gate.py:80]: > list(reversed_permutation_map.keys())[0]
this line is O(n)

Comment (NoureldinYosri) [reply cirq-core/cirq/ops/permutation_gate.py:80]: you can make your implementation O(n) by changing that line to 

```py3
a = next(iter(reversed_permutation_map))
```

but even in that case the implementation would be a bit convoluted

Comment (NoureldinYosri) [reply cirq-core/cirq/ops/permutation_gate.py:80]: @migueltorrescosta `==`

Comment (NoureldinYosri) [reply cirq-core/cirq/ops/permutation_gate.py]: can you commit your code?

Comment (NoureldinYosri) [reply cirq-core/cirq/ops/permutation_gate.py]: all the CI checks pass for current commits, where is this test failing?
image

Comment (migueltorrescosta) [reply cirq-core/cirq/ops/permutation_gate.py]: The tests fail when I attempted this suggested solution, https://github.com/quantumlib/Cirq/pull/6588#discussion_r1593149835

Comment (migueltorrescosta) [reply cirq-core/cirq/ops/permutation_gate.py]: @NoureldinYosri I believe this is ready to go as is. Do mention it if there is something I should change.

Comment (NoureldinYosri) [reply cirq-core/cirq/ops/permutation_gate.py]: please commit the code that fails

File: cirq-core/cirq/ops/permutation_gate.py [hunk 2]
```diff
@@ -74,23 +75,19 @@ def _has_unitary_(self):
...
-                is_sorted = False
-
-        while not is_sorted:
-            is_sorted = True
-            for i in range(0, n - 1, 2):
-                yield from _swap_if_out_of_order(i)
-            for i in range(1, n - 1, 2):
-                yield from _swap_if_out_of_order(i)
+        # List is need otherwise permutation[j] = -1 fails to assign to a tuple
+        permutation = list(deepcopy(self.permutation))
```
Comment (NoureldinYosri) [inline cirq-core/cirq/ops/permutation_gate.py:79]: nit:
```py3
permutation = [p for p in self.permutation]
```

**Review by NoureldinYosri — CHANGES_REQUESTED** [review]

**Comment by migueltorrescosta** [comment]
Suggestions applied. The shorter code feels easier to follow:
```python
remaining_permutations = {v: i for i, v in enumerate(self.permutation)}

while remaining_permutations:
    a = next(iter(remaining_permutations))
    b = remaining_permutations.pop(a)
    while b in remaining_permutations.keys():
        yield swap_gates.SWAP(qubits[a], qubits[b])
        (a, b) = (b, remaining_permutations.pop(b))
```
vs
```python
permutation = [p for p in self.permutation]

for i in range(len(permutation)):

    if permutation[i] == -1:
        continue
    cycle = [i]
    while permutation[cycle[-1]] != i:
        cycle.append(permutation[cycle[-1]])

    for j in cycle:
        permutation[j] = -1

    for idx in cycle[1:]:
        yield swap_gates.SWAP(qubits[cycle[0]], qubits[idx])
```

It might be a question of personal taste, so I am happy to have either merged

**Review by NoureldinYosri — APPROVED** [review]
thanks for your help with this task

fixes https://github.com/quantumlib/Cirq/issues/5097

**Pull request merged**